pkg/operation/botanist/secrets.go

```diff
@@ -16,925 +16,852 @@ package botanist
-apiVersion: v1
-kind: Config
-current-context: {{.ClusterName}}
-clusters:
-- name: {{.ClusterName}}
-  cluster:
-    certificate-authority-data: {{.CACertificate}}
-    server: https://{{.APIServerURL}}
-contexts:
-- name: {{.ClusterName}}
-  context:
-    cluster: {{.ClusterName}}
-    user: {{.ClusterName}}
-users:
-- name: {{.ClusterName}}
-  user:
-    client-certificate-data: {{.ClientCertificate}}
-    client-key-data: {{.ClientKey}}
-{{- if and (gt (len .BasicAuthUsername) 0) (gt (len .BasicAuthPassword) 0)}}
-- name: {{.ClusterName}}-basic-auth
-  user:
-    username: {{.BasicAuthUsername}}
-    password: {{.BasicAuthPassword}}
-{{- end}}`
+// DeploySecrets creates a CA certificate for the Shoot cluster and uses it to sign the server certificate
+// used by the kube-apiserver, and all client certificates used for communcation. It also creates RSA key
+// pairs for SSH connections to the nodes/VMs and for the VPN tunnel. Moreover, basic authentication
+// credentials are computed which will be used to secure the Ingress resources and the kube-apiserver itself.
+// Server certificates for the exposed monitoring endpoints (via Ingress) are generated as well.
+func (b *Botanist) DeploySecrets() error {
+	var (
+		certificateAuthorities = map[string]*secrets.Certificate{}
+		existingSecretsMap     = map[string]*corev1.Secret{}
+
+		secretMapLock               = sync.RWMutex{}
+		certificateAuthorityMapLock = sync.RWMutex{}
+
+		wg        sync.WaitGroup
+		errorList []error
+	)
+
+	// Read all existing secrets to only re-generate those which do not yet exist.
+	secretList, err := b.K8sSeedClient.ListSecrets(b.Shoot.SeedNamespace, metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+	for _, secret := range secretList.Items {
+		secretObj := secret
+		existingSecretsMap[secret.Name] = &secretObj
+	}
+
+	// Previously, we have used the same certificate authority in all places. Now, we are using dedicated CAs for the
+	// different components. Gardener does not re-create/generate certificates/secrets if they already exist. Hence,
+	// we have to delete the existing certificates to let them re-created and re-signed by a new CA.
+	var dedicatedCASecrets = map[string][]string{
+		caETCD:       []string{"etcd-server-tls", "etcd-client-tls"},
+		caFrontProxy: []string{"kube-aggregator"},
+		caKubelet:    []string{"kube-apiserver-kubelet"},
+	}
+	for caName, secrets := range dedicatedCASecrets {
+		for _, secretName := range secrets {
+			if _, ok := existingSecretsMap[caName]; !ok {
+				if err := b.K8sSeedClient.DeleteSecret(b.Shoot.SeedNamespace, secretName); err != nil && !apierrors.IsNotFound(err) {
+					return err
+				}
+				delete(existingSecretsMap, secretName)
+			}
+		}
+	}
+
+	//////////////////////////////////////////////////////////////////////////////////////////////
+	// Generation of certificate authorities in order to sign other server or client certificates.
+	for name, config := range wantedCertificateAuthorities {
```

Comment:
This pipeline inline is quite big - can we refactor it so it has smaller parts + enhanced error reporting?
I'd rather vote for a pipeline as in https://blog.golang.org/pipelines